src/except.cpp

```diff
@@ -82,40 +82,106 @@ std::string to_string(const Tag tag) {
-         "}, {";
+    s += to_string(dims.labels()[i], separator) + ", " +
+         std::to_string(dims.shape()[i]) + "}, {";
   s.resize(s.size() - 3);
-  s += '}';
+  s += "}\n";
   return s;
 }
 
-std::string to_string(const Unit &unit) {
-  switch (unit.id()) {
-  case Unit::Id::Dimensionless:
-    return "Unit::Dimensionless";
-  case Unit::Id::Length:
-    return "Unit::Length";
+std::string to_string(const DType dtype) {
+  switch (dtype) {
+  case DType::String:
+    return "string";
+  case DType::Bool:
+    return "bool";
+  case DType::Char:
+    return "char";
+  case DType::Dataset:
+    return "dataset";
+  case DType::Float:
+    return "float";
+  case DType::Double:
+    return "double";
+  case DType::Int32:
+    return "int32";
+  case DType::Int64:
+    return "int64";
+  case DType::Unknown:
+    return "unknown";
   default:
-    return "<unknown unit>";
+    return "unregistered dtype";
+  };
+}
+
+std::string to_string(const Unit &unit, const std::string separator) {
+  return std::regex_replace(do_to_string(unit), std::regex("::"), separator);
+}
+std::string to_string(const Dim dim, const std::string separator) {
+  return std::regex_replace(do_to_string(dim), std::regex("::"), separator);
+}
+
+std::string to_string(const Tag tag, const std::string separator) {
+  return std::regex_replace(do_to_string(tag), std::regex("::"), separator);
+}
+
+// For use with variables
+std::string make_dims_labels(const Variable &variable,
+                             const std::string separator) {
+  auto dims = variable.dimensions();
+  std::string diminfo = "( ";
+  for (int32_t i = 0; i < dims.ndim(); ++i) {
+    diminfo += to_string(dims.labels()[i], separator);
+    if (i != dims.ndim() - 1) {
+      diminfo += ", ";
+    }
   }
+  diminfo += " )";
+  return diminfo;
 }
 
-std::string to_string(const Variable &variable) {
-  std::string s("Variable(");
-  s += to_string(variable.tag()) + ", " + variable.name() + ")";
+std::string to_string(const Variable &variable, const std::string separator) {
+  std::string variableName = variable.name();
+  std::string diminfo = make_dims_labels(variable, separator);
+  if (variableName.empty())
+    variableName = "''";
+  std::string s = "Variable(";
+  s += to_string(variable.tag(), separator) + ", " + variableName + "," +
+       diminfo + ", " + to_string(variable.dtype()) + ")\n";
   return s;
-}
+} // namespace dataset
 
-std::string to_string(const Dataset &dataset) {
+std::string to_string(const Dataset &dataset, const std::string separator) {
   std::string s("Dataset with ");
   s += std::to_string(dataset.size()) + " variables";
+  s += "\nDimensions : " + to_string(dataset.dimensions(), separator);
+  // The following is peformed to allow variables to be sorted into catagories
+  // of coordinate, data and attribute as part of output
+  for (const auto &var : dataset) {
+    s += to_string(var, separator);
```

Comment:
Where did the `isCoord` etc. disappear? You are now suddenly mixing coords and data? Also, the comment is outdated?

Reply:
Yes, oversight on comment. Output now looks like this for example:
```
Dataset with 6 variables
Dimensions : {{Dim.Y, 10}, {Dim.X, 100}}
Variable(Coord.X, '',( Dim.X ), int64)
Variable(<unknown tag>, '',( Dim.Y ), int64)
Variable(Data.Value, a,( Dim.X, Dim.Y ), double)
Variable(Data.Variance, a,( Dim.X, Dim.Y ), double)
Variable(Data.Value, b,( Dim.X, Dim.Y ), double)
Variable(Data.Variance, b,( Dim.X, Dim.Y ), double)
```
I removed the sorting in line with the earlier discussion. I also removed the prefix `Coord`, `Data`.. from the strings as it appears in the tag anyway.

Reply:
The earlier discussion suggested to remove sorting with the coord/data category. We should still list coordinates in their own section (see the code snippet I had provided).